File: backend/src/application/use-cases/categorization/ProductCategorizationUseCase.ts
import { ProductCategory } from "@domain/entities/IProduct";
import { ProductNameNormalizer } from "@domain/services/ProductNameNormalizer";
import type { KeywordCategorizer } from "@application/ports/outgoing/KeywordCategorizer";
import type { AiCategorizer } from "@application/ports/outgoing/AiCategorizer";
import type { CategoryCacheRepository } from "@application/ports/outgoing/CategoryCacheRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";

export interface ProductCategorizationUseCaseDeps {
	keywordCategorizer: KeywordCategorizer;
	aiCategorizer?: AiCategorizer;
	categoryCache?: CategoryCacheRepository;
	aiTimeoutMs?: number;
	logger?: LoggerPort;
}

export class ProductCategorizationUseCase {
	private readonly keywordCategorizer: KeywordCategorizer;
	private readonly aiCategorizer?: AiCategorizer;
	private readonly categoryCache?: CategoryCacheRepository;
	private readonly aiTimeoutMs: number;
	private readonly logger?: LoggerPort;

	constructor(deps: ProductCategorizationUseCaseDeps) {
		this.keywordCategorizer = deps.keywordCategorizer;
		this.aiCategorizer = deps.aiCategorizer;
		this.categoryCache = deps.categoryCache;
		this.aiTimeoutMs = deps.aiTimeoutMs ?? 1500;
		this.logger = deps.logger;
	}

	async execute(productName: string): Promise<ProductCategory> {
		const normalizedName = ProductNameNormalizer.normalize(productName);

		const cachedCategory = await this.lookupInCache(normalizedName);
		if (cachedCategory !== undefined) {
			return cachedCategory;
		}

		const keywordCategory = this.keywordCategorizer.match(productName);
		if (keywordCategory !== undefined) {
			await this.writeThroughToCache(normalizedName, keywordCategory);
			return keywordCategory;
		}

		if (!this.aiCategorizer) {
			return ProductCategory.OTHER;
		}

		try {
			const aiCategory = await this.withTimeout(
				this.aiCategorizer.categorize(productName),
				this.aiTimeoutMs,
			);
			if (aiCategory !== undefined) {
				await this.writeThroughToCache(normalizedName, aiCategory);
				return aiCategory;
			}
			return ProductCategory.OTHER;
		} catch (error) {
			this.logger?.error(
				`[ProductCategorizationUseCase] AI fallback failed: ${String(error)}`,
			);
			return ProductCategory.OTHER;
		}
	}

	private async lookupInCache(
		normalizedName: string,
	): Promise<ProductCategory | undefined> {
		if (!this.categoryCache) {
			return undefined;
		}

		try {
			const cached = await this.categoryCache.findByNormalizedName(normalizedName);
			if (cached === undefined) {
				return undefined;
			}

			const matchedCategory = Object.values(ProductCategory).find(
				(cat) => cat === cached,
			);
			return matchedCategory;
		} catch (error) {
			this.logger?.warn(
				`[ProductCategorizationUseCase] Cache lookup failed, falling through: ${String(error)}`,
			);
			return undefined;
		}
	}

	private async writeThroughToCache(
		normalizedName: string,
		category: ProductCategory,
	): Promise<void> {
		if (!this.categoryCache) {
			return;
		}

		try {
			await this.categoryCache.upsertMany([{ normalizedName, category }]);
		} catch (error) {
			this.logger?.warn(
				`[ProductCategorizationUseCase] Cache write-through failed: ${String(error)}`,
			);
		}
	}

	private async withTimeout<T>(
		promise: Promise<T>,
		timeoutMs: number,
	): Promise<T> {
		let timer: NodeJS.Timeout | undefined;

		try {
			return await Promise.race([
				promise,
				new Promise<T>((_resolve, reject) => {
					timer = setTimeout(() => {
						reject(new Error(`AI categorization timeout after ${timeoutMs}ms`));
					}, timeoutMs);
				}),
			]);
		} finally {
			if (timer) {
				clearTimeout(timer);
			}
		}
	}
}

File: backend-dotnet/src/api/Application/UseCases/Categorization/ProductCategorizationUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Models;
using SupermarketPlanner.Api.Domain.Services;

namespace SupermarketPlanner.Api.Application.UseCases.Categorization;

public sealed record ProductCategorizationOptions(int AiTimeoutMs);

public sealed class ProductCategorizationUseCase(
    IKeywordCategorizer keywordCategorizer,
    IAiCategorizer? aiCategorizer,
    ICategoryCacheRepository? categoryCache,
    ILogger<ProductCategorizationUseCase>? logger,
    ProductCategorizationOptions options)
{
    public async Task<Either<DomainError, string>> Invoke(string productName, CancellationToken cancellationToken)
    {
        var normalizedName = ProductNameNormalizer.Normalize(productName);

        var cachedCategory = await LookupInCache(normalizedName, cancellationToken);
        if (cachedCategory is not null)
            return Either<DomainError, string>.FromRight(cachedCategory);

        var keywordCategory = keywordCategorizer.Match(productName);
        if (keywordCategory is ProductCategory category)
        {
            await WriteThroughToCache(normalizedName, category.ToString(), cancellationToken);
            return Either<DomainError, string>.FromRight(category.ToString());
        }

        return Either<DomainError, string>.FromRight(
            await GetAiCategory(productName, normalizedName, cancellationToken));
    }

    private async Task<string> GetAiCategory(string productName,
        string normalizedName, CancellationToken cancellationToken)
    {
        if (aiCategorizer is null)
            return nameof(ProductCategory.Other);

        try
        {
            var aiCategory = await RunWithTimeout(
                aiCategorizer.Categorize(productName, cancellationToken),
                options.AiTimeoutMs,
                cancellationToken);

            if (aiCategory is not ProductCategory resolvedCategory)
            {
                logger?.LogWarning(
                    "[ProductCategorizationUseCase] AI categorization timed out after {TimeoutMs}ms",
                    options.AiTimeoutMs);
                return nameof(ProductCategory.Other);
            }

            await WriteThroughToCache(normalizedName, resolvedCategory.ToString(), cancellationToken);
            return resolvedCategory.ToString();
        }
        catch (Exception ex)
        {
            logger?.LogError(
                "[ProductCategorizationUseCase] AI fallback failed: {Error}",
                ex.Message);
            return nameof(ProductCategory.Other);
        }
    }

    private async Task<string?> LookupInCache(string normalizedName, CancellationToken cancellationToken)
    {
        if (categoryCache is null)
            return null;

        try
        {
            var result = await categoryCache.FindByNormalizedName(normalizedName, cancellationToken);
            return result.Match<string?>(onSome: category => category, onNone: () => null);
        }
        catch (Exception exception)
        {
            logger?.LogWarning(
                "[ProductCategorizationUseCase] Cache lookup failed, falling through: {Error}",
                exception.Message);
            return null;
        }
    }

    private async Task WriteThroughToCache(string normalizedName, string category, CancellationToken cancellationToken)
    {
        if (categoryCache is null)
            return;

        (await categoryCache.UpsertMany(
            [new CategoryCacheEntry(normalizedName, category)],
            cancellationToken))
            .Match(
                onLeft: error =>
                {
                    logger?.LogWarning(
                        "[ProductCategorizationUseCase] Cache write-through failed: {Error}",
                        error.Message);
                    return Unit.Value;
                },
                onRight: _ => Unit.Value);
    }

    private static async Task<ProductCategory?> RunWithTimeout(
        Task<ProductCategory?> task, int timeoutMs, CancellationToken cancellationToken)
    {
        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutCts.CancelAfter(timeoutMs);
        var completedTask = await Task.WhenAny(task, Task.Delay(timeoutMs, timeoutCts.Token));
        if (completedTask != task) return null;
        await timeoutCts.CancelAsync();
        return await task;
    }
}
